aiohttp/client.py

```diff
@@ -218,6 +227,18 @@ def _request(self, method, url, *,
         handle = tm.start()
 
         url = URL(url)
+
+        if trace_context is None:
+            trace_context = SimpleNamespace()
```

Comment:
I think the API is wrong here.
User will never send a `trace_context` into `async with session.get()` explicitly.
It's another level of abstraction.
What the user will do is setting up session properly on initialization stage by substribing on signals and (optionally) providing a *trace context factory* for creating a new container for user data.
I even doubt if we need a factory parameter, at least at current stage.

Reply:
The rationale behind this implementation is the following one:

Give the proper freedom to the developer to have a grain control of his requests calls building trace context for each request.

Perhaps

```python
async def on_request_start(trace_context, host, port, headers):
    trace_context['start'] = loop.time()

async def on_request_end(trace_context, resp):
    await send_metrics(
        time=loop.time() - trace_context['start']
        query=trace_context['query']
    )


sesion = ClientSession()
session.add_on_requests_start(on_request_start)
session.add_on_requests_end(on_request_end)

resp = session.get("http://localhost?query=foo", trace={'query':'foo'})
resp = session.get("http://localhost?query=foo", trace={'query':'bar'})
```

This example shows how the same ClientSession is used to make different queries that might have divergent traces. 

In case the user is keen on share information between all requests that belong to the same `ClientSession` might use a closure pattern, perhaps:

```python
async def on_request_end(query):
    async def on_request_end(trace_context, resp):
        await send_metrics(
            time=loop.time() - trace_context['start']
            query=query
        )
    return on_request_end

sesion = ClientSession()
session.add_on_requests_start(on_request_start)
session.add_on_requests_end(on_request_end(query='foo')
```

I can see that your point about forcing the user to populate each request call can be less kindy, but from my experience have the way to pass a context that has information about the current execution is a must. Also, take into account that having this granularity allows the user to implement Session or Requests context.

Another solution would pass for implement two different contexts, one for the session and another one for the request. But IMHO overcomplicates right now the implementation.

Thoughts ?